### PR TITLE
fix(xaml): Ensure generator idempotence

### PR DESCRIPTION
**Pull request by jeromelaban**
## PR Type

What kind of change does this PR introduce?
- Bugfix

## What is the new behavior?

Adjusts XAML generation to be idempotent in the context of metadata generation.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/GenerationInfoManager.cs [hunk 1]
```diff
@@ -18,21 +22,74 @@ internal class GenerationRunInfoManager
...
+				context,
+				out var useXamlReaderHotReload,
+				out var useHotReload);
+
+			var hash = context
+				.AdditionalFiles
+				.Aggregate(0, (hash, f) => ByteSequenceComparer.GetHashCode(f.GetText()?.GetChecksum() ?? ImmutableArray<byte>.Empty) ^ hash);
+
+			// Only create a new run when the previous run additional files are different
+			// This ensures that each run produces the same output for a given input.
```
Comment (agneszitte-nventive) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/GenerationInfoManager.cs:55]:
```suggestion
			// Only create a new run when the previous run's additional files are different.
			// This ensures that each run produces the same output for a given input.
```

**Comment by Youssef1313** [comment]
FYI @davidwengier :)

File: src/SourceGenerators/Uno.UI.SourceGenerators/Content/Uno.UI.SourceGenerators.props [hunk 2]
```diff
@@ -44,9 +44,6 @@
 									)
 								  )">false</UnoForceHotReloadCodeGen>
 
-		<!-- Use Application.LoadComponent for HotReload scenarios -->
-		<UnoUseXamlReaderHotReload>true</UnoUseXamlReaderHotReload>
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/Content/Uno.UI.SourceGenerators.props:48]: Is this change related to the issue we are trying to solve?

File: src/SourceGenerators/Uno.UI.SourceGenerators/Uno.UI.SourceGenerators.csproj [hunk 3]
```diff
@@ -91,10 +91,13 @@
...
 		<PropertyGroup>
 			<_baseNugetPath Condition="'$(USERPROFILE)'!=''">$(USERPROFILE)</_baseNugetPath>
-      <_baseNugetPath Condition="'$(HOME)'!=''">$(HOME)</_baseNugetPath>
+			<_baseNugetPath Condition="'$(HOME)'!=''">$(HOME)</_baseNugetPath>
+			
+			<_OverridePackageId>Uno.UI</_OverridePackageId>
+			<_OverridePackageId Condition="'$(UNO_UWP_BUILD)'=='false'">Uno.WinUI</_OverridePackageId>
 
-			<_TargetNugetFolder Condition="'$(TargetFramework)'=='net461'">$(_baseNugetPath)\.nuget\packages\Uno.UI\$(UnoNugetOverrideVersion)\tools\uno-sourcegen</_TargetNugetFolder>
-			<_TargetNugetFolder Condition="'$(TargetFramework)'=='netstandard2.0'">$(_baseNugetPath)\.nuget\packages\Uno.UI\$(UnoNugetOverrideVersion)\analyzers\dotnet\cs</_TargetNugetFolder>
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/Uno.UI.SourceGenerators.csproj:97]: I guess we have the same issue for other projects (e.g, Lottie, Svg)

https://github.com/unoplatform/uno/blob/a6eda7c5d1ab8d2b34cfc443fbd54e9432b8cf16/src/AddIns/Uno.UI.Svg/Uno.UI.Svg.csproj#L53

https://github.com/unoplatform/uno/blob/a6eda7c5d1ab8d2b34cfc443fbd54e9432b8cf16/src/AddIns/Uno.UI.Lottie/Uno.UI.Lottie.csproj#L54

But that can be fixed in a separate PR for sure.

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/GenerationInfoManager.cs [hunk 4]
```diff
@@ -18,21 +22,74 @@ internal class GenerationRunInfoManager
...
+		private static void ReadProjectConfiguration(GeneratorExecutionContext context, out bool useXamlReaderHotReload, out bool useHotReload)
+		{
+			bool.TryParse(context.GetMSBuildPropertyValue("UnoUseXamlReaderHotReload"), out useXamlReaderHotReload);
+
+			var configuration = context.GetMSBuildPropertyValue("Configuration")
+				?? throw new InvalidOperationException("The configuration property must be provided");
+
+			var isDebug = string.Equals(configuration, "Debug", StringComparison.OrdinalIgnoreCase);
+
+			useHotReload = true;
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/GenerationInfoManager.cs:84]: We provide a value both in the `if` and `else`. This should be redundant.

```suggestion
```

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/GenerationInfoManager.cs [hunk 5]
```diff
@@ -18,21 +22,74 @@ internal class GenerationRunInfoManager
...
 		}
 
+		private static void ReadProjectConfiguration(GeneratorExecutionContext context, out bool useXamlReaderHotReload, out bool useHotReload)
+		{
+			bool.TryParse(context.GetMSBuildPropertyValue("UnoUseXamlReaderHotReload"), out useXamlReaderHotReload);
+
+			var configuration = context.GetMSBuildPropertyValue("Configuration")
+				?? throw new InvalidOperationException("The configuration property must be provided");
+
+			var isDebug = string.Equals(configuration, "Debug", StringComparison.OrdinalIgnoreCase);
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/GenerationInfoManager.cs:82]: Consider moving to the `else` below where these are actually used.

File: src/SourceGenerators/Uno.UI.SourceGenerators/Content/Uno.UI.SourceGenerators.props (hunk 2 above)
Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/Content/Uno.UI.SourceGenerators.props:48]: It is, indeed.

File: src/SourceGenerators/Uno.UI.SourceGenerators/Uno.UI.SourceGenerators.csproj (hunk 3 above)
Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/Uno.UI.SourceGenerators.csproj:97]: Yes, we need to adjust most packages to use the same top-level props file.

**Review by carldebilly — APPROVED** [review]
